File: backend/src/routes/schedules.js
import express from 'express';
import { pool } from '../db/connection.js';

const router = express.Router();

// Get all schedules
router.get('/', async (req, res) => {
  try {
    const { route_type, direction, date } = req.query;
    
    let query = `
      SELECT 
        ss.*,
        s.name as shuttle_name,
        s.capacity,
        c.name as company_name,
        COUNT(sr.id) as registered_count
      FROM shuttle_schedules ss
      JOIN shuttles s ON ss.shuttle_id = s.id
      JOIN companies c ON s.company_id = c.id
      LEFT JOIN shuttle_registrations sr ON ss.id = sr.schedule_id
        AND sr.registration_date = COALESCE($1, CURRENT_DATE)
        AND sr.status = 'confirmed'
      WHERE ss.is_active = true
    `;
    
    const params = [date || null];
    let paramCount = 1;
    
    if (route_type) {
      paramCount++;
      query += ` AND ss.route_type = $${paramCount}`;
      params.push(route_type);
    }
    
    if (direction) {
      paramCount++;
      query += ` AND ss.direction = $${paramCount}`;
      params.push(direction);
    }
    
    query += ` GROUP BY ss.id, s.id, c.id ORDER BY ss.departure_time`;
    
    const result = await pool.query(query, params);
    res.json(result.rows);
  } catch (error) {
    console.error('Error fetching schedules:', error);
    res.status(500).json({ error: 'Failed to fetch schedules' });
  }
});

// Get schedule by ID
router.get('/:id', async (req, res) => {
  try {
    const result = await pool.query(
      `SELECT 
        ss.*,
        s.name as shuttle_name,
        s.capacity,
        c.name as company_name
      FROM shuttle_schedules ss
      JOIN shuttles s ON ss.shuttle_id = s.id
      JOIN companies c ON s.company_id = c.id
      WHERE ss.id = $1`,
      [req.params.id]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Schedule not found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error fetching schedule:', error);
    res.status(500).json({ error: 'Failed to fetch schedule' });
  }
});

// Create new schedule
router.post('/', async (req, res) => {
  try {
    const {
      shuttle_id,
      route_type,
      direction,
      departure_time,
      arrival_time,
      days_of_week
    } = req.body;
    
    const result = await pool.query(
      `INSERT INTO shuttle_schedules 
      (shuttle_id, route_type, direction, departure_time, arrival_time, days_of_week)
      VALUES ($1, $2, $3, $4, $5, $6)
      RETURNING *`,
      [shuttle_id, route_type, direction, departure_time, arrival_time, days_of_week || [1,2,3,4,5]]
    );
    
    res.status(201).json(result.rows[0]);
  } catch (error) {
    console.error('Error creating schedule:', error);
    res.status(500).json({ error: 'Failed to create schedule' });
  }
});

// Update schedule
router.put('/:id', async (req, res) => {
  try {
    const {
      departure_time,
      arrival_time,
      days_of_week,
      is_active
    } = req.body;
    
    const result = await pool.query(
      `UPDATE shuttle_schedules 
      SET departure_time = COALESCE($1, departure_time),
          arrival_time = COALESCE($2, arrival_time),
          days_of_week = COALESCE($3, days_of_week),
          is_active = COALESCE($4, is_active)
      WHERE id = $5
      RETURNING *`,
      [departure_time, arrival_time, days_of_week, is_active, req.params.id]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Schedule not found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error updating schedule:', error);
    res.status(500).json({ error: 'Failed to update schedule' });
  }
});

// Delete schedule
router.delete('/:id', async (req, res) => {
  try {
    const result = await pool.query(
      'DELETE FROM shuttle_schedules WHERE id = $1 RETURNING id',
      [req.params.id]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Schedule not found' });
    }
    
    res.json({ message: 'Schedule deleted successfully' });
  } catch (error) {
    console.error('Error deleting schedule:', error);
    res.status(500).json({ error: 'Failed to delete schedule' });
  }
});

// Get schedules organized by route and direction for frontend display
router.get('/organized/display', async (req, res) => {
  try {
    const { date } = req.query;
    
    const query = `
      SELECT 
        ss.route_type,
        ss.direction,
        ss.departure_time,
        ss.arrival_time,
        s.name as shuttle_name,
        s.capacity,
        c.name as company_name,
        COUNT(sr.id) as registered_count
      FROM shuttle_schedules ss
      JOIN shuttles s ON ss.shuttle_id = s.id
      LEFT JOIN companies c ON s.company_id = c.id
      LEFT JOIN shuttle_registrations sr ON ss.id = sr.schedule_id
        AND sr.registration_date = COALESCE($1, CURRENT_DATE)
        AND sr.status = 'confirmed'
      WHERE ss.is_active = true
      GROUP BY ss.id, ss.route_type, ss.direction, ss.departure_time, ss.arrival_time, s.name, s.capacity, c.name
      ORDER BY ss.route_type, ss.direction, ss.departure_time
    `;
    
    const result = await pool.query(query, [date || null]);
    
    // Organize data by route type and direction
    const organized = {
      savidor_to_tzafrir: {
        outbound: [],
        return: []
      },
      kiryat_aryeh_to_tzafrir: {
        outbound: [],
        return: []
      }
    };
    
    result.rows.forEach(row => {
      // Format time for display (remove seconds)
      const displayTime = row.departure_time.substring(0, 5);
      
      const scheduleEntry = {
        time: displayTime,
        fullTime: row.departure_time,
        arrivalTime: row.arrival_time ? row.arrival_time.substring(0, 5) : null,
        shuttleName: row.shuttle_name,
        capacity: row.capacity,
        companyName: row.company_name,
        registeredCount: parseInt(row.registered_count) || 0
      };
      
      if (organized[row.route_type] && organized[row.route_type][row.direction]) {
        organized[row.route_type][row.direction].push(scheduleEntry);
      }
    });
    
    res.json(organized);
  } catch (error) {
    console.error('Error fetching organized schedules:', error);
    res.status(500).json({ error: 'Failed to fetch organized schedules' });
  }
});

export default router;

File: backend/app/routers/schedules.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, delete
from typing import List
from uuid import UUID

from ..database import get_database_session
from ..models import ShuttleSchedule, Shuttle, Company
from ..logging_manager import logger_manager
from ..schemas import ShuttleSchedule as ScheduleSchema, ScheduleCreate, ScheduleUpdate, MessageResponse, OrganizedSchedules, ScheduleEntry, RouteSchedules
from ..auth import get_current_active_user, AdminUser

router = APIRouter()

@router.get("/public", response_model=List[ScheduleSchema])
async def get_schedules_public(
    db: AsyncSession = Depends(get_database_session)
):
    result = await db.execute(
        select(ShuttleSchedule).order_by(ShuttleSchedule.departure_time)
    )
    schedules = result.scalars().all()
    return schedules

@router.get("/", response_model=List[ScheduleSchema])
async def get_schedules(
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    logger_manager.info("Fetching schedules list", {"user_id": str(current_user.id)})
    
    result = await db.execute(
        select(ShuttleSchedule).order_by(ShuttleSchedule.departure_time)
    )
    schedules = result.scalars().all()
    
    logger_manager.info("Schedules retrieved", {"count": len(schedules)})
    return schedules

@router.get("/{schedule_id}", response_model=ScheduleSchema)
async def get_schedule(
    schedule_id: UUID,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    result = await db.execute(
        select(ShuttleSchedule).where(ShuttleSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()
    
    if not schedule:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Schedule not found"
        )
    
    return schedule

@router.post("/", response_model=ScheduleSchema)
async def create_schedule(
    schedule_data: ScheduleCreate,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    logger_manager.info("Creating new schedule", {
        "shuttle_id": str(schedule_data.shuttle_id),
        "departure_time": str(schedule_data.departure_time),
        "user_id": str(current_user.id)
    })
    # Verify shuttle exists
    result = await db.execute(
        select(Shuttle).where(Shuttle.id == schedule_data.shuttle_id)
    )
    shuttle = result.scalar_one_or_none()
    
    if not shuttle:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Shuttle not found"
        )
    
    new_schedule = ShuttleSchedule(
        shuttle_id=schedule_data.shuttle_id,
        route_type=schedule_data.route_type,
        direction=schedule_data.direction,
        departure_time=schedule_data.departure_time,
        arrival_time=schedule_data.arrival_time,
        days_of_week=schedule_data.days_of_week,
        is_active=schedule_data.is_active
    )
    
    db.add(new_schedule)
    await db.commit()
    await db.refresh(new_schedule)
    
    return new_schedule

@router.put("/{schedule_id}", response_model=ScheduleSchema)
async def update_schedule(
    schedule_id: UUID,
    schedule_data: ScheduleUpdate,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    # Check if schedule exists
    result = await db.execute(
        select(ShuttleSchedule).where(ShuttleSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()
    
    if not schedule:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Schedule not found"
        )
    
    # Update only provided fields
    update_data = schedule_data.dict(exclude_unset=True)
    if update_data:
        await db.execute(
            update(ShuttleSchedule)
            .where(ShuttleSchedule.id == schedule_id)
            .values(**update_data)
        )
        await db.commit()
        await db.refresh(schedule)
    
    return schedule

@router.delete("/{schedule_id}", response_model=MessageResponse)
async def delete_schedule(
    schedule_id: UUID,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    # Check if schedule exists
    result = await db.execute(
        select(ShuttleSchedule).where(ShuttleSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()
    
    if not schedule:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Schedule not found"
        )
    
    # Delete schedule
    await db.execute(
        delete(ShuttleSchedule).where(ShuttleSchedule.id == schedule_id)
    )
    await db.commit()
    
    return MessageResponse(message="Schedule deleted successfully")

@router.get("/shuttle/{shuttle_id}", response_model=List[ScheduleSchema])
async def get_schedules_by_shuttle(
    shuttle_id: UUID,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    result = await db.execute(
        select(ShuttleSchedule)
        .where(ShuttleSchedule.shuttle_id == shuttle_id)
        .order_by(ShuttleSchedule.departure_time)
    )
    schedules = result.scalars().all()
    return schedules

@router.get("/organized/display/public", response_model=OrganizedSchedules)
async def get_organized_schedules_public(
    date: str = None,
    db: AsyncSession = Depends(get_database_session)
):
    return await _get_organized_schedules(db)

@router.get("/organized/display", response_model=OrganizedSchedules)
async def get_organized_schedules(
    date: str = None,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    return await _get_organized_schedules(db)

async def _get_organized_schedules(db: AsyncSession) -> OrganizedSchedules:
    # Get all schedules with shuttle and company info
    result = await db.execute(
        select(ShuttleSchedule, Shuttle.name.label('shuttle_name'), Shuttle.capacity, Company.name.label('company_name'))
        .join(Shuttle, ShuttleSchedule.shuttle_id == Shuttle.id)
        .join(Company, Shuttle.company_id == Company.id)
        .where(ShuttleSchedule.is_active == True)
        .order_by(ShuttleSchedule.departure_time)
    )
    
    schedules_data = result.all()
    
    # Organize schedules by route and direction
    organized = {
        "savidor_to_tzafrir": {"outbound": [], "return": []},
        "kiryat_aryeh_to_tzafrir": {"outbound": [], "return": []}
    }
    
    for schedule, shuttle_name, capacity, company_name in schedules_data:
        # Format time
        departure_time = schedule.departure_time
        time_str = departure_time.strftime("%H:%M")
        full_time_str = departure_time.strftime("%H:%M:%S")
        
        arrival_time_str = None
        if schedule.arrival_time:
            arrival_time_str = schedule.arrival_time.strftime("%H:%M")
        
        schedule_entry = ScheduleEntry(
            time=time_str,
            fullTime=full_time_str,
            arrivalTime=arrival_time_str,
            shuttleName=shuttle_name,
            capacity=capacity,
            companyName=company_name,
            registeredCount=0  # TODO: Calculate actual registration count
        )
        
        # Add to appropriate route and direction
        if schedule.route_type in organized:
            organized[schedule.route_type][schedule.direction].append(schedule_entry)
    
    return OrganizedSchedules(
        savidor_to_tzafrir=RouteSchedules(
            outbound=organized["savidor_to_tzafrir"]["outbound"],
            return_=organized["savidor_to_tzafrir"]["return"]
        ),
        kiryat_aryeh_to_tzafrir=RouteSchedules(
            outbound=organized["kiryat_aryeh_to_tzafrir"]["outbound"],
            return_=organized["kiryat_aryeh_to_tzafrir"]["return"]
        )
    )